# Two-Qubit Hadamard Statevector and Normalization

## 1. Problem Statement
Construct a 2-qubit quantum circuit initialized in $|00\rangle$, apply Hadamard gates to both qubits, obtain the resulting statevector, display the amplitudes of the computational basis states, and verify that the sum of the corresponding probabilities is equal to 1.

---

## 2. Theoretical Background & Concepts

### Initial Ground State
In quantum computing and Qiskit, multi-qubit registers are initialized by default in the all-zero computational basis state:

$$|00\rangle = |0\rangle \otimes |0\rangle = \begin{pmatrix} 1 \\ 0 \\ 0 \\ 0 \end{pmatrix}$$

### Equal 2-Qubit Superposition
The **Hadamard gate** ($H$) creates an equal linear superposition of computational basis states. When applied independently to both qubits of a system initialized in $|00\rangle$, the combined unitary operation is the tensor product $H \otimes H = H^{\otimes 2}$:

$$|\psi\rangle = (H \otimes H)|00\rangle = (H|0\rangle) \otimes (H|0\rangle)$$

Substituting the single-qubit transformation $H|0\rangle = \frac{|0\rangle + |1\rangle}{\sqrt{2}}$:

$$|\psi\rangle = \left(\frac{|0\rangle + |1\rangle}{\sqrt{2}}\right) \otimes \left(\frac{|0\rangle + |1\rangle}{\sqrt{2}}\right) = \frac{1}{2}(|00\rangle + |01\rangle + |10\rangle + |11\rangle)$$

This produces an equal superposition across all four computational basis states.

### Statevector and Amplitudes
A **statevector** $|\psi\rangle$ is a complex unit vector in a $2^n$-dimensional Hilbert space that completely characterizes the pure state of an $n$-qubit system:

$$|\psi\rangle = \sum_{i=0}^{2^n-1} \alpha_i |i\rangle = \begin{pmatrix} \alpha_0 \\ \alpha_1 \\ \vdots \\ \alpha_{2^n-1} \end{pmatrix}$$

where each $\alpha_i \in \mathbb{C}$ represents the probability amplitude associated with the computational basis state $|i\rangle$.

For this 2-qubit circuit:
- Each computational basis state ($|00\rangle, |01\rangle, |10\rangle, |11\rangle$) has an identical amplitude:
  $$\alpha_i = \frac{1}{2} = 0.5$$
- According to the **Born rule**, the probability of measuring basis state $|i\rangle$ upon wavefunction collapse is the squared magnitude of its amplitude:
  $$P(i) = |\alpha_i|^2 = \left|\frac{1}{2}\right|^2 = \frac{1}{4} = 0.25 \quad (25\%)$$
  for each state $i \in \{00, 01, 10, 11\}$.

### Statevector Normalization
A foundational postulate of quantum mechanics requires that the total probability of all mutually exclusive measurement outcomes must equal 1:

$$\sum_i |\alpha_i|^2 = 1$$

where $\alpha_i$ are the statevector amplitudes. Verifying that the sum of the squared magnitudes equals 1 confirms that the statevector is properly normalized and represents a physically valid quantum state.

---

## 3. Circuit Flow
The transformation sequence is:

$$|00\rangle \longrightarrow H \text{ on q0 and } H \text{ on q1} \longrightarrow \text{equal 2-qubit superposition} \longrightarrow \text{statevector} \longrightarrow \text{normalization check}$$

---

## 4. Tools & Libraries Used
- **Qiskit**: IBM's open-source quantum computing framework for constructing and visualizing quantum circuits (`QuantumCircuit`).
- **`qiskit.quantum_info.Statevector`**: Analytical module for extracting exact quantum statevectors directly from circuit instructions without measurement collapse.
- **NumPy**: Numerical computation library used for vector operations and verifying normalization ($\sum |\alpha_i|^2$).

### Step 1: Import Required Libraries
Import `QuantumCircuit` and `Statevector` from Qiskit, and `numpy` for probability calculations.

In [1]:
from qiskit import QuantumCircuit
from qiskit.quantum_info import Statevector
import numpy as np


### Step 2: Construct the 2-Qubit Circuit
Initialize a 2-qubit quantum circuit. By convention, both qubits start in the default computational ground state $|00\rangle$.

In [2]:
# Create circuit
qc = QuantumCircuit(2)


### Step 3: Apply Hadamard Gates
Apply a Hadamard gate to qubit 0 (`qc.h(0)`) and qubit 1 (`qc.h(1)`) to place both qubits into an equal superposition.

In [3]:
# Apply Hadamard to both qubits
qc.h(0)
qc.h(1)


### Step 4: Obtain the Statevector
Extract the exact quantum statevector from the circuit instructions using `Statevector.from_instruction(qc)`, and display the circuit diagram and statevector.

In [4]:
# Get statevector
state = Statevector.from_instruction(qc)

print("Quantum Circuit:")
print(qc)

print("\nStatevector:")
print(state)


Quantum Circuit:
     ┌───┐
q_0: ┤ H ├
     ├───┤
q_1: ┤ H ├
     └───┘

Statevector:
Statevector([0.5+0.j, 0.5+0.j, 0.5+0.j, 0.5+0.j],
            dims=(2, 2))


### Step 5: Display Computational-Basis Amplitudes
Iterate over the statevector amplitudes (`state.data`) and format each basis state with its 2-bit binary label (`|00>`, `|01>`, `|10>`, `|11>`).

In [5]:
# Display amplitudes
print("\nAmplitudes:")
for i, amp in enumerate(state.data):
    print(f"|{i:02b}> : {amp}")



Amplitudes:
|00> : (0.4999999999999999+0j)
|01> : (0.4999999999999999+0j)
|10> : (0.4999999999999999+0j)
|11> : (0.4999999999999999+0j)


### Step 6: Verify Statevector Normalization
Compute the sum of the squared absolute magnitudes of all state amplitudes using `np.sum(np.abs(state.data)**2)` to verify probability conservation.

In [6]:
# Verify normalization
probability = np.sum(np.abs(state.data)**2)

print("\nSum of probabilities =", probability)



Sum of probabilities = 0.9999999999999996


## Conclusion

- **Equal 2-Qubit Superposition**: Applying Hadamard gates to both qubits transforms the initial ground state $|00\rangle$ into an equal superposition across all four computational basis states:
  $$|\psi\rangle = \frac{1}{2}(|00\rangle + |01\rangle + |10\rangle + |11\rangle)$$
- **Equal Amplitudes & Probabilities**: Each computational basis state ($|00\rangle, |01\rangle, |10\rangle, |11\rangle$) possesses an identical probability amplitude of $\alpha_i \approx 0.5$, corresponding to an equal theoretical measurement probability of $P = |\alpha_i|^2 = 0.25$ ($25\%$).
- **Normalization Confirmed**: The normalization calculation yields $\sum |\alpha_i|^2 = 0.9999999999999996 \approx 1.0$, mathematically confirming that total probability is conserved and that the resulting statevector represents a physically valid, normalized quantum state.